sdk/program/src/nonce/state/mod.rs

```diff
@@ -2,21 +2,59 @@
 
 mod current;
 pub use current::{Data, DurableNonce, State};
-use serde_derive::{Deserialize, Serialize};
+use {
+    crate::hash::Hash,
+    serde_derive::{Deserialize, Serialize},
+};
 
 #[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
 pub enum Versions {
+    Legacy(Box<State>),
     Current(Box<State>),
 }
 
 impl Versions {
-    pub fn new_current(state: State) -> Self {
-        Self::Current(Box::new(state))
+    pub fn new(state: State, separate_domains: bool) -> Self {
+        if separate_domains {
+            Self::Current(Box::new(state))
+        } else {
+            Self::Legacy(Box::new(state))
+        }
     }
 
-    pub fn convert_to_current(self) -> State {
+    pub fn state(&self) -> &State {
         match self {
-            Self::Current(state) => *state,
+            Self::Legacy(state) => state,
+            Self::Current(state) => state,
+        }
+    }
+
+    pub fn separate_domains(&self) -> bool {
+        match self {
+            Self::Legacy(_) => false,
+            Self::Current(_) => true,
+        }
+    }
+
+    // Returns true if the recent_blockhash field in Transaction verifies.
+    pub fn verify_recent_blockhash(&self, hash: &Hash) -> bool {
```

Comment:
The behavior here should be influenced by feature activations because legacy nonces can still be used on devnet and testnet until `separate_nonce_from_blockhash` is activated

Reply:
done